File: src/services/cmd.ts
import { invoke } from '@tauri-apps/api/core';
import { ConfigSchema } from '@/schemas/config';
import type {
  Config,
  McpServerConfig,
  ShellType,
  FrontendConfig,
  LogLevel,
} from '@/schemas/config';

export async function greet(): Promise<void> {
  return invoke<void>('greet');
}

const LOG_LEVEL_PRIORITY: Record<LogLevel, number> = {
  trace: 0,
  debug: 1,
  info: 2,
  warn: 3,
  error: 4,
};

let _currentLogLevel: LogLevel = 'trace';

export function setLogLevelFilter(level: LogLevel) {
  _currentLogLevel = level;
}

export async function logMessage(level: LogLevel, message: string): Promise<void> {
  if (LOG_LEVEL_PRIORITY[level] < LOG_LEVEL_PRIORITY[_currentLogLevel]) return;
  return invoke<void>('log_message', { level, message });
}

export async function setLogLevel(level: LogLevel): Promise<void> {
  return invoke<void>('set_log_level', { level });
}

export type { Config, McpServerConfig, ShellType, FrontendConfig };

// Config commands
export async function getConfig(): Promise<Config> {
  const data = await invoke<unknown>('get_config');
  return ConfigSchema.parse(data);
}

export async function updateConfig(newConfig: Config): Promise<void> {
  return invoke<void>('update_config', { newConfig });
}

// MCP types
export interface ToolInfo {
  server_name: string;
  tool: {
    name: string;
    description?: string;
    inputSchema: Record<string, unknown>;
  };
}

export interface TextContent {
  type: 'text';
  text: string;
}

export interface ImageContent {
  type: 'image';
  /** mcp:// URL after materialization */
  data: string;
  mimeType: string;
}

export interface EmbeddedResource {
  type: 'resource';
  resource: {
    uri: string;
    text?: string;
    blob?: string;
    mimeType?: string;
  };
}

export type ContentBlock = TextContent | ImageContent | EmbeddedResource;

export interface CallToolResult {
  content: ContentBlock[];
  structuredContent?: unknown;
  isError?: boolean;
  _meta?: Record<string, unknown>;
}

export type ServerStatus = 'Starting' | 'Running' | 'Stopped' | { Failed: { error: string } };

// MCP commands
export async function mcpListTools(): Promise<ToolInfo[]> {
  return invoke<ToolInfo[]>('mcp_list_tools');
}

export async function mcpCallTool(
  serverName: string,
  toolName: string,
  arguments_?: Record<string, unknown>
): Promise<CallToolResult> {
  return invoke<CallToolResult>('mcp_call_tool', {
    serverName,
    toolName,
    arguments: arguments_,
  });
}

export async function mcpServerStatus(serverName: string): Promise<ServerStatus> {
  return invoke<ServerStatus>('mcp_server_status', { serverName });
}

// MCP Panel types
export interface ServerInfo {
  name: string;
  version: string;
  has_tools: boolean;
  has_prompts: boolean;
  has_resources: boolean;
}

export interface PromptInfo {
  server_name: string;
  prompt: {
    name: string;
    description?: string;
    arguments?: Array<{
      name: string;
      description?: string;
      required?: boolean;
    }>;
  };
}

export interface ResourceInfo {
  server_name: string;
  resource: {
    uri: string;
    name: string;
    description?: string;
    mime_type?: string;
  };
}

// MCP content store
export async function mcpStoreContent(path: string): Promise<string> {
  return invoke<string>('mcp_store_content', { path });
}

export async function mcpStoreImageBytes(data: number[], mimeType: string): Promise<string> {
  return invoke<string>('mcp_store_image_bytes', { data, mimeType });
}

// MCP Panel commands
export async function mcpServerInfo(serverName: string): Promise<ServerInfo | null> {
  return invoke<ServerInfo | null>('mcp_server_info', { serverName });
}

export async function mcpListPrompts(): Promise<PromptInfo[]> {
  return invoke<PromptInfo[]>('mcp_list_prompts');
}

export async function mcpListResources(): Promise<ResourceInfo[]> {
  return invoke<ResourceInfo[]>('mcp_list_resources');
}

File: src-tauri/src/cmd.rs
use crate::kernel::context::AppContext;
use crate::MCP_MANAGER;
use logging::{logging, Type};
use mcp::{CallToolResult, ServerStatus, ToolInfo};

#[tauri::command]
pub fn greet() {
    logging!(info, Type::Cmd, "greet called, sending ping");
    AppContext::send_ping();
}

#[tauri::command]
pub fn log_message(level: String, message: String) {
    match level.as_str() {
        "error" => logging!(error, Type::Frontend, "{}", message),
        "warn" => logging!(warn, Type::Frontend, "{}", message),
        "debug" => logging!(debug, Type::Frontend, "{}", message),
        "trace" => logging!(trace, Type::Frontend, "{}", message),
        _ => logging!(info, Type::Frontend, "{}", message),
    }
}

#[tauri::command]
pub async fn mcp_list_tools() -> Result<Vec<ToolInfo>, String> {
    logging!(info, Type::Cmd, "mcp_list_tools called");
    let manager = MCP_MANAGER.get().ok_or("MCP manager not initialized")?;
    Ok(manager.list_tools().await)
}

#[tauri::command]
pub async fn mcp_call_tool(
    server_name: String,
    tool_name: String,
    arguments: Option<serde_json::Value>,
) -> Result<CallToolResult, String> {
    logging!(
        info,
        Type::Cmd,
        "mcp_call_tool called: server={}, tool={}",
        server_name,
        tool_name
    );
    let manager = MCP_MANAGER.get().ok_or("MCP manager not initialized")?;
    manager
        .call_tool(&server_name, &tool_name, arguments)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn mcp_server_status(server_name: String) -> Result<ServerStatus, String> {
    logging!(info, Type::Cmd, "mcp_server_status called: server={}", server_name);
    let manager = MCP_MANAGER.get().ok_or("MCP manager not initialized")?;
    manager
        .server_status(&server_name)
        .await
        .map_err(|e| e.to_string())
}
